src/coreComponents/managers/FieldSpecification/FieldSpecificationOps.hpp

```diff
@@ -478,10 +478,22 @@ struct FieldSpecificationEqual : public FieldSpecificationOp< OpEqual >
       localIndex const numEntries = matrix.numNonZeros( localRow );
 
       real64 diagonal = 0;
+      real64 const minDiagonal = 1e-15;
       for( localIndex j = 0; j < numEntries; ++j )
       {
         if( columns[ j ] == dof )
-        { diagonal = entries[ j ]; }
+        {
+          // check that the entry is large enough to enforce the boundary condition
+          if( entries[j] >= 0 && entries[j] < minDiagonal )
+          {
+            entries[ j ] = minDiagonal;
+          }
+          else if( entries[j] < 0 && entries[j] > -minDiagonal )
+          {
+            entries[ j ] = -minDiagonal;
+          }
+          diagonal = entries[j];
+        }
```

Comment:
@klevzoff One thing that did not work well with the analytical derivatives of the new Dead-Oil model was in this function: with analytical Dead-Oil the derivatives of the mass conservation equation of component `ic` wrt the component density of `jc` (`jc \neq ic`) are extremely small (~ 1e-18), and for the first iteration of the first time step, they are simply equal to zero (`dR_ic/dRho_jc = 0`). So, in the equation/variable ordering of the compositional solver (below), `entries[dof]` is equal to zero. As a result, when we apply the Dirichlet BC, the matrix becomes singular because we zero out the other entries on this row. 

I added a check for the value of `entries[dof]`, and that fixes the issue. However, I did not do anything for the other `SpecifyFieldValue` function (using `LAI::ParallelMatrix`) just above because it is not entirely clear to me that we still use it (and where?) since we moved to `CRSMatrix` for the assembly.

Ordering for the compositional solver
```
                  dPres dRho_{ic = 0} dRho_{ic = 1} dRho_{ic = 2}
Mass Cons ic = 0
Mass Cons ic = 1
Mass Cons ic = 2
Vol Balance
```

Reply:
@francoishamon I don't think we use the other version of `SpecifyFieldValue` (in fact, it needs to be removed), so this fix should be just fine.

As for the equation/variable alignment, fixing this has been on the radar since the beginning but I never got around to implementing it. The plan was to have the first equation be the total mass conservation (which would align with pressure DoF), followed by component mass cons up from ic=0 to NC-2; and also have the last DoF be total density instead of last component's density (which would align with vol balance). This ordering would guarantee nonzero diagonals and would probably lead to slightly more stable simulations.

I was initially hoping to achieve this with kernel-usable dense LA approach (pre- and post-multiplying each cell's local jacobian by small matrices to rearrange rows and columns into other linear combinations), which we didn't have at the time; now we could try something with `tensorOps`. However, there' a lot of zeros in those permutation matrices, so a hand-rolled reordering routine might be better.
 
The equation lineup isn't too difficult to change by hand at the final stage of the kernel where we assemble component mass balances. Changing the last DOF is a more annoying task. Both changes will propagate to other kernels (residual norm, apply/check solution, etc). Reservoir-well solvers would be affected as well.

Reply:
Ok I see, this seems to be an equation/variable alignment that would be nice to have. 

Regarding the change in equation lineup, it would be really useful to have the choice for the first equation between 1) total mass conservation, and 2) the pressure equation of the sequential implicit method. If I understood correctly the way the SI pressure equation is typically constructed, we need to compute weights (and their derivatives wrt pressure) that are used to form a linear combination of the mass conservation equations that eliminates the derivatives wrt to the non-pressure dofs. I think this needs to happen at the nonlinear level, since the derivatives of the weights wrt pressure need to be introduced (but maybe there is a smarter way to do it entirely at the linear level, I am not sure yet). I am saying all this because @jafranc will be implementing his research on the SI method in GEOSX, and this capability will be very useful to have at that point.

Just to make sure I understand, when you suggest the `tensorOps` way and/or the hand-rolled reordering, would you do that A) after each `compute` function (accumulation, flux, wells, etc) when we call the RAJA atomic adds, or B) just once when the `CRSMatrix` is fully assembled and it is time to apply the boundary conditions? If the answer is B), maybe we could try this first, less invasive step before addressing the change in the last dof?

Reply:
I was initially thinking about A (specifically because a dense pre/post-multiply would be really easy and fast to do when you have a local dense block on the thread stack). However B sounds quite feasible for re-arranging equations, in particular because we can apply it to entire assembled rows as you mention, so it works effortlessly for coupled solvers. The downside is we need to do an extra kernel launch and operate on global memory again, but it's probably not too bad. This approach also seems less convenient for forming SI pressure equation because you need to manipulate individual row entries (pressure derivatives) in CSR as opposed to just adding entire rows (which you can do blindly, provided that sparsity patterns are the same between rows). But we should definitely explore this idea.